Compiler Suggests Cloning an `async` Block

### Code

```Rust
// https://github.com/HTGAzureX1212/HarTex/blob/491fe72da9042d5a5c7f61d0d5cdaf6e9b112057/discord-frontend/hartex-discord-leader/src/kafka.rs#L79-L120
```


### Current output

```Shell
error[E0382]: use of moved value: `producer`
   --> hartex-discord-leader\src\kafka.rs:78:23
    |
68  |   async fn inbound<'a, Q>(shards: Vec<&'a mut Shard<Q>>, producer: FutureProducer) -> miette::Result<()>
    |                                                          -------- move occurs because `producer` has type `FutureProducer`, which does not implement the `Copy` trait
...
74  |       loop {
    |       ---- inside of this loop
...
77  |           for shard in shards {
    |           ------------------- inside of this loop
78  |               set.spawn(async move {
    |  _______________________^
79  | |                 while let Some(result) = shard.next().await {
80  | |                     match result {
81  | |                         Ok(message) => {
...   |
95  | |                             if let Err((error, _)) = producer
    | |                                                      -------- use occurs due to use in coroutine
...   |
119 | |                 }
120 | |             });
    | |_____________^ value moved here, in previous iteration of loop
    |
help: consider cloning the value if the performance cost is acceptable
    |
120 |             }.clone());
    |              ++++++++
```


### Desired output

```Shell
error[E0382]: use of moved value: `producer`
   --> hartex-discord-leader\src\kafka.rs:78:23
    |
68  |   async fn inbound<'a, Q>(shards: Vec<&'a mut Shard<Q>>, producer: FutureProducer) -> miette::Result<()>
    |                                                          -------- move occurs because `producer` has type `FutureProducer`, which does not implement the `Copy` trait
...
74  |       loop {
    |       ---- inside of this loop
...
77  |           for shard in shards {
    |           ------------------- inside of this loop
78  |               set.spawn(async move {
    |  _______________________^
79  | |                 while let Some(result) = shard.next().await {
80  | |                     match result {
81  | |                         Ok(message) => {
...   |
95  | |                             if let Err((error, _)) = producer
    | |                                                      -------- use occurs due to use in coroutine
...   |
119 | |                 }
120 | |             });
    | |_____________^ value moved here, in previous iteration of loop
    |
help: consider cloning the value if the performance cost is acceptable
    |
95  |                               if let Err((error, _)) = producer.clone()
    |                                                                ++++++++
```


### Rationale and extra context

I expect that the `.clone()` invocation to be added for the value that is moved, but for some reason it is doing it for the entire `async` closure.

### Other cases

_No response_

### Rust Version

```Shell
rustc 1.78.0-nightly (8f359beca 2024-02-23)
binary: rustc
commit-hash: 8f359beca4e58bc3ae795a666301a8f47023044c
commit-date: 2024-02-23
host: x86_64-pc-windows-msvc
release: 1.78.0-nightly
LLVM version: 18.1.0
```


### Anything else?

_No response_

--- a/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
@@ -631,7 +631,7 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                     }
 
                     // FIXME: We make sure that this is a normal top-level binding,
-                    // but we could suggest `todo!()` for all uninitalized bindings in the pattern pattern
+                    // but we could suggest `todo!()` for all uninitialized bindings in the pattern pattern
                     if let hir::StmtKind::Let(hir::LetStmt { span, ty, init: None, pat, .. }) =
                         &ex.kind
                         && let hir::PatKind::Binding(..) = pat.kind
@@ -759,7 +759,7 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
         true
     }
 
-    /// In a move error that occurs on a call wihtin a loop, we try to identify cases where cloning
+    /// In a move error that occurs on a call within a loop, we try to identify cases where cloning
     /// the value would lead to a logic error. We infer these cases by seeing if the moved value is
     /// part of the logic to break the loop, either through an explicit `break` or if the expression
     /// is part of a `while let`.
@@ -960,7 +960,7 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
             {
                 // FIXME: We could check that the call's *parent* takes `&mut val` to make the
                 // suggestion more targeted to the `mk_iter(val).next()` case. Maybe do that only to
-                // check for wheter to suggest `let value` or `let mut value`.
+                // check for whether to suggest `let value` or `let mut value`.
 
                 let span = in_loop.span;
                 if !finder.found_breaks.is_empty()

--- a/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs
@@ -4,6 +4,7 @@
 #![allow(rustc::untranslatable_diagnostic)]
 
 use either::Either;
+use hir::ClosureKind;
 use rustc_data_structures::captures::Captures;
 use rustc_data_structures::fx::FxIndexSet;
 use rustc_errors::{codes::*, struct_span_code_err, Applicability, Diag, MultiSpan};
@@ -463,6 +464,15 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                 } else if let UseSpans::FnSelfUse { kind: CallKind::Normal { .. }, .. } = move_spans
                 {
                     // We already suggest cloning for these cases in `explain_captures`.
+                } else if let UseSpans::ClosureUse {
+                    closure_kind:
+                        ClosureKind::Coroutine(CoroutineKind::Desugared(_, CoroutineSource::Block)),
+                    args_span: _,
+                    capture_kind_span: _,
+                    path_span,
+                } = move_spans
+                {
+                    self.suggest_cloning(err, ty, expr, path_span);
                 } else if self.suggest_hoisting_call_outside_loop(err, expr) {
                     // The place where the the type moves would be misleading to suggest clone.
                     // #121466

--- a/tests/ui/borrowck/cloning-in-async-block-121547.rs
+++ b/tests/ui/borrowck/cloning-in-async-block-121547.rs
@@ -1,0 +1,11 @@
+//@ edition:2021
+
+async fn clone_async_block(value: String) {
+    for _ in 0..10 {
+        async { //~ ERROR: use of moved value: `value` [E0382]
+            drop(value);
+            //~^ HELP: consider cloning the value if the performance cost is acceptable
+        }.await
+    }
+}
+fn main() {}

--- a/tests/ui/borrowck/cloning-in-async-block-121547.stderr
+++ b/tests/ui/borrowck/cloning-in-async-block-121547.stderr
@@ -1,0 +1,22 @@
+error[E0382]: use of moved value: `value`
+  --> $DIR/cloning-in-async-block-121547.rs:5:9
+   |
+LL |   async fn clone_async_block(value: String) {
+   |                              ----- move occurs because `value` has type `String`, which does not implement the `Copy` trait
+LL |       for _ in 0..10 {
+   |       -------------- inside of this loop
+LL | /         async {
+LL | |             drop(value);
+   | |                  ----- use occurs due to use in coroutine
+LL | |
+LL | |         }.await
+   | |_________^ value moved here, in previous iteration of loop
+   |
+help: consider cloning the value if the performance cost is acceptable
+   |
+LL |             drop(value.clone());
+   |                       ++++++++
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0382`.

